pkg/alertmanager/operator.go

```diff
@@ -469,6 +637,336 @@ func (c *Operator) sync(ctx context.Context, key string) error {
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list namespaces")
+	}
+	return ns, nil
+}
+
+func (c *Operator) createOrUpdateTLSAssetSecret(am *monitoringv1.Alertmanager) error {
+	boolTrue := true
+	sClient := c.kclient.CoreV1().Secrets(am.Namespace)
+
+	tlsAssets, err := c.loadTLSAssets(am)
+	if err != nil {
+		return err
+	}
+
+	tlsAssetsSecret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   tlsAssetsSecretName(am.Name),
+			Labels: c.config.Labels.Merge(managedByOperatorLabels),
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion:         am.APIVersion,
+					BlockOwnerDeletion: &boolTrue,
+					Controller:         &boolTrue,
+					Kind:               am.Kind,
+					Name:               am.Name,
+					UID:                am.UID,
+				},
+			},
+		},
+		Data: make(map[string][]byte, len(tlsAssets)),
+	}
+
+	for key, asset := range tlsAssets {
+		tlsAssetsSecret.Data[key] = []byte(asset)
+	}
+
+	_, err = sClient.Get(context.TODO(), tlsAssetsSecret.Name, metav1.GetOptions{})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return errors.Wrapf(
+				err,
+				"failed to check whether tls assets secret already exists for Alertmanager %v in namespace %v",
+				am.Name,
+				am.Namespace,
+			)
+		}
+		_, err = sClient.Create(context.TODO(), tlsAssetsSecret, metav1.CreateOptions{})
+		level.Debug(c.logger).Log("msg", "created tlsAssetsSecret", "secretname", tlsAssetsSecret.Name)
+
+	} else {
+		_, err = sClient.Update(context.TODO(), tlsAssetsSecret, metav1.UpdateOptions{})
+		level.Debug(c.logger).Log("msg", "updated tlsAssetsSecret", "secretname", tlsAssetsSecret.Name)
+	}
+
+	if err != nil {
+		return errors.Wrapf(err, "failed to create TLS assets secret for Alertmanager %v in namespace %v", am.Name, am.Namespace)
+	}
+
+	return nil
+}
+
+// TLSAsset represents any tls related opaque string, e.g. ca files, client
+// certificates.
+type TLSAsset string
+
+func (c *Operator) loadTLSAssets(am *monitoringv1.Alertmanager) (map[string]TLSAsset, error) {
+
+	amcAssets := map[string]TLSAsset{}
+	assets := map[string]TLSAsset{}
+
+	amcReceiverTLSConfigs := []*monitoringv1.TLSConfig{}
+
+	amcs, err := c.selectAlertManagerConfigs(am)
+	if err != nil {
+		return nil, errors.Wrap(err, "selecting AlertManagerConfigs failed")
+	}
+
+	for _, amc := range amcs {
+		for _, receiver := range amc.Spec.Receivers {
+			for _, pdr := range receiver.PagerDutyConfigs {
+				if pdr.HTTPConfig != nil && pdr.HTTPConfig.TLSConfig != nil {
+					amcReceiverTLSConfigs = append(amcReceiverTLSConfigs, pdr.HTTPConfig.TLSConfig)
+				}
+			}
+		}
+		amcAssets, err = c.loadTLSAssetsFromTLSConfigList(amc.Namespace, amcReceiverTLSConfigs)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	for k, v := range amcAssets {
+		assets[k] = v
+	}
+
+	return assets, nil
+}
+
+func (c *Operator) loadTLSAssetsFromTLSConfigList(ns string, tlsConfigs []*monitoringv1.TLSConfig) (map[string]TLSAsset, error) {
```

Comment:
One thing that I'm not sure about -- the `monitoringv1.TLSConfig` type has additional fields that aren't handled here: `CAFile, CertFile, KeyFile`. I don't think they're read by the prometheus controller either. Is there something we should be doing with those?

Reply:
`CAFile`, `CertFile` and `KeyFile` aren't things that we want to support for new objects because it isn't flexible enough for users compared to key selectors (see [comment](https://github.com/prometheus-operator/prometheus-operator/pull/3370#discussion_r461665236)).

Reply:
Got it :+1: For now I'll stop using that `monitoringv1.TLSConfig` and make an equivalent to the `SafeTLSConfig` type and maybe once #3370 is merged we can switch to that.

Reply:
Done.